Bugfix badge truncation

vijkris99 opened the pull request:
- The main update is a bugfix for the test on the description and test badges getting truncated
- I also updated the badges to read the last build's status, instead of the last successful build's status

vramas-icims committed:
Fixed badge length bug

* Fixed the issue with badge text getting truncated with the build
description and test status badges
* Still needs some minor adjustments to the padding logic :)

vramas-icims committed:
Adjusted the badge padding calculations

vramas-icims committed:
Using lastBuild() instead of lastSuccessfulBuild() for all badges

vramas-icims committed:
Ignoring failing tests

vramas-icims committed:
Using lastCompletedBuild() instead of lastBuild() for all badges

SxMShaDoW commented on src/main/java/org/jenkinsci/plugins/badge/PublicBadgeAction.java:
```diff
@@ -127,11 +128,11 @@ public HttpResponse doCoverageIcon(StaplerRequest req, StaplerResponse rsp, @Que
         Job<?, ?> project = getProject(job);
         Integer codeCoverage = null;
 
-        if (project.getLastSuccessfulBuild() != null) {
```
So the problem is if you get the lastCompleted build, if it was aborted/failed then the code coverage would show it as 0 (most likely the case if a CI job fails).

vijkris99 replied:
Right, that is what I intended. The idea is to correlate all badges with the build status badge, so it is a true representation of the latest status.
Maybe a good middle ground is to have this behavior configurable through Jenkins?

SxMShaDoW replied:
Well I don't know if the latest failed build accurately reflects the code coverage of the repo. Its the latest successful build that does.

vijkris99 replied:
Agreed! My view was that the code coverage could have changed since the last successful build, and it may be better to show something like "n/a" instead of a stale value. I updated the code coverage badge to default to "n/a" instead of "0" when a build fails or aborts. If you feel this behavior is not right, please feel free to change back. However, I do believe we should still use the last completed build for all other badges regardless.

SxMShaDoW commented on src/main/java/org/jenkinsci/plugins/badge/StatusImage.java:
```diff
@@ -94,11 +96,11 @@
     @Override
     public void generateResponse(StaplerRequest req, StaplerResponse rsp, Object node) throws IOException, ServletException {
-        String v = req.getHeader("If-None-Match");
+        /*String v = req.getHeader("If-None-Match");
```
this was taken from the forked repo. Is this no longer needed?

vijkris99 replied:
I think so. When I was changing the badge length logic, this piece of code was continuing to serve the older version of the badge. That made it super difficult to debug.

SxMShaDoW replied:
We may want to be careful. There was caching built into this repo to prevent a call to Jenkins every time.

vijkris99 replied:
This piece of code can be re-enabled safely once this version stabilizes. However, we should document that this piece of code should be disabled whenever someone attempts to modify the code, otherwise, none of the changes would get reflected. I have left it commented out for now.

SxMShaDoW commented on src/main/java/org/jenkinsci/plugins/badge/StatusImage.java:
```diff
@@ -34,7 +35,7 @@
      * TO DO
      */
-    private final byte[] payload;
+    public final byte[] payload;
```
why did we make this public?

vijkris99 replied:
My bad! I was testing something out, and forgot to undo it. It should be private!

SxMShaDoW commented on src/test/java/org/jenkinsci/plugins/badge/PublicBadgeActionTest.java:
```diff
@@ -174,13 +177,15 @@ public void validAnonymousViewStatusAccess() throws Exception {
      */
     @PresetData(ANONYMOUS_READONLY)
     @Test
+    @Ignore
```
can you explain why we are ignoring this?

vijkris99 replied:
To be honest, I was not sure how the test was working. I noticed that all tests related to authentication were failing. I figured it may have something to do with updates to Jenkins, and the tests becoming stale. Do you know what this test does?

SxMShaDoW replied:
Authentication probably changed since you bumped it to 2.X. I would recommend keeping it at 1.6 unless you plan to change the tests to reflect the new auth changes.

SxMShaDoW commented on src/test/java/org/jenkinsci/plugins/badge/PublicBadgeActionTest.java:
```diff
@@ -150,7 +153,7 @@ public void invalidAnonymousAccess() throws Exception {
      * @throws Exception
      */
-    @Test
+    /*@Test
```
can you explain this please :)

vijkris99 replied:
Line 161 inside this test method throws a compilation error on my machine, and I didn't have the time to troubleshoot. I am not sure if this happens only on my machine.

SxMShaDoW commented:
@vijkris99 - Hey. Thanks for the contribution! I added a few comments and noticed that travis is failing. Could you look into this?

vijkris99 commented:
I had errors with the POM file on my machine. I am using Eclipse, and some of it is due to known issues as mentioned in https://wiki.jenkins.io/display/JENKINS/Plugin+tutorial#Plugintutorial-Eclipse. I modified some of the dependencies to solve some of these issues, and this may be the reason for Travis failing.

vramas-icims committed:
Attempting to fix failed Travis build

* Rolled back dependencies to their original versions mentioned in the
POM file
* "Un"-ignored failing tests
* Updated code coverage badge to display "n/a" instead of "0" on
failed/aborted builds

vijkris99 commented:
I think I fixed the Travis build now. There is a marginal drop in code coverage, but I hope that's ok. Let me know what you think.

vijkris99 commented:
@SxMShaDoW Did you get a chance to review this PR? Pls let me know if you have any questions, thanks!

SxMShaDoW commented:
Hi @vijkris99 - I don't feel comfortable merging this. 

I think if we made it a configuration either at the system level or on the embeddable badges page I would be more okay with it. 

On top of that, I know this repo doesn't have a lot of code coverage, but I don't want to drop it even lower. 

Lastly, commenting out that etag with no proposed solution makes me hesitant. I think the caching is critical when your system has 100s of jobs. Maybe putting that in a config may be plausible too.

vramas-icims committed:
Fixed caching issue

* Included the plugin's current version number as part of the generated
ETag, so that previously cached badges are discarded whenever the plugin
is updated
* Updated code coverage default value to "NaN" instead of "n/a"
* Bumped up POM version number

vramas-icims committed:
Reverted to lastSuccessfulBuild() for the code coverage badge

sputnikci commented on src/main/java/org/jenkinsci/plugins/badge/PublicBadgeAction.java:
```diff
@@ -148,20 +149,19 @@ public HttpResponse doCoverageIcon(StaplerRequest req, StaplerResponse rsp, @Que
+            if (cloverInstalled != null && cloverInstalled.isActive()) {
+                CloverBuildAction cloverAction = project.getLastSuccessfulBuild().getAction(CloverBuildAction.class);
+                if (cloverAction != null){
+                    if (cloverAction.getBuildHealth() != null){
```
\[PMD\] INFO: These nested if statements could be combined

vijkris99 commented:
@SxMShaDoW I fixed the caching issue by including the plugin's current version as part of the ETag. I also reverted the code coverage badge to use the last successful build instead. Let me know if this works.

SxMShaDoW commented:
great job!

SxMShaDoW commented:
note: I don't have any infrastructure to actually publish the `.hpi` you generate to a central jenkins / nexus. I can look into that if you want.

vijkris99 commented:
@SxMShaDoW That would be great! I was thinking that we should try hosting this plugin on the Jenkins community repo, as described @ https://wiki.jenkins.io/display/JENKINS/Hosting+Plugins. I am sure there are many users who would be interested in using this.
